### DISCUSSION TITLE
mcrun & typescript woes

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1276] [Q&A]
I'm having a problem with typescript modules and mcrun. The problem is how to reference typescript modules found in the host without double-loading them into the mod. In detail:

- the host is written in typescript, it defines a module A
- the mod is written in typescript, it imports A
- the tsc compiler needs the definition of A when compiling the mod
- but the module A should not be included in the mod binary since it's already loaded in the host
- to satisfy tsc the module A needs to be in the manifest's include but that automatically includes it into the mod archive

If the host is written in javascript, then the manifest of the mod can point to an A.d.ts file, which satisfies tsc. So right now the only solution I see is to create a redundant A.d.ts file for my module A.ts? It seems what is needed is a "typescript_include" section in the mod's manifest for modules that need to go into the tsconfig.json but not into the final archive. I have the feeling I'm missing some obvious solution?

**Comment by tve** [comment]
So I somewhat get to answer my question...

The first step is to make tsc output type definitions for the host. This can be accomplished by adding to the host's manifest:
```json
  "typescript": {
    "tsconfig": {
      "compilerOptions": {
        "declaration": true,
        "declarationDir": "../../../../../../../../types"
      }
    }
  }
```
This causes the build dir to have a directory tree with the type definitions.

The second step is to reference those type definitions in the module, again using the manifest:
```json
  "typescript": {
    "tsconfig": {
      "compilerOptions": {
        "paths": {
          "embedded:network/http/rest-server": [
            "/home/src/moddable/js-board/host/build/types/modules/lib/rest-server/rest-server"
          ]
        },
        "rootDirs": [".", "../host/build/types"]
      }
    }
  }
```

The rootDirs option adds the types tree from step 1 into the build of the mod. However, the module called "embedded:network/http/rest-server" is in that tree at "modules/lib/rest-server" so I had to create an explicit mapping using the paths. Oddly the '../host' reference worked in rootDirs but a similar relative reference doesn't work in paths...

So the remaining issue here is that the module name mappings captured in various manifests needs to be recreated manually. This is something mcrun could help with...

**Reply by phoddie** [reply]
Maybe I'm missing something, but why have tsc output declarations for the host and use those, rather than adding the needed `.d.ts` source file to the manifest?

**Reply by tve** [reply]
The host is written in typescript so there are no `.d.ts` source files. But given where I ended up I can actually import the source `.ts` files in the tsconfig since mcrun won't see those and hence won't load them into the mod archive. So the typescript section in the host manifest can be removed and the typescript section in the mod manifest looks as below. The remaining ugly thing is that the module name mappings need to be constructed manually, I'll probably write a script that post-processes the tsconfig produced for the host...
Thanks for the nudge that made me see the simpler solution!

```json
{
  "build": {
    "TVEMOD": "/home/src/moddable/modules"
  },
  "include": [
    "$(MODDABLE)/examples/manifest_mod.json",
    "$(MODDABLE)/examples/manifest_typings.json"
  ],
  "modules": {
    "*": ["./*.js", "./*.ts"]
  },
  "typescript": {
    "tsconfig": {
      "compilerOptions": {
        "paths": {
          "embedded:network/http/rest-server": ["$(TVEMOD)/lib/rest-server/rest-server"],
          "sys": ["$(TVEMOD)/lib/sys/sys"]
        },
        "rootDirs": [".", "../host"]
      }
    }
  }
}
```

Update: the `rootDirs` option doesn't work because it causes files to be deposited elsewhere by tsc...

**Reply by phoddie** [reply]
Ah, so since you have a TypeScript implementation you don't need separate `.d.ts` files. I missed that.

Looking at that from a manifest perspective... you can't include the `.ts` sources in the modules section, as they would be compiled into the output. Having another section -- say `"declarations"` -- could help there, perhaps? That could build the `tsconfig` bits for you, without building any `.ts` sources referenced. Not sure how practical that would be to implement.

**Reply by tve** [reply]
Yup! However, let me first try to post-process the tsconfig generated for the host. Basically the host gets built. It includes a pile of modules. It then makes sense to produce a `manifest-export.json` for the host that has the relevant `typescript>tsconfig>compilerOptions>paths` to use those modules. With that generated, all the mods can `include` that manifest just like including other manifests for libraries that do get compiled in.

**Reply by phoddie** [reply]
Sounds good. I'm definitely not ready to hump into manifest changes. Just trying to think through what it might look like as a way to confirm my understanding.

**Reply by tve** [reply]
Well, this is way more wicked than I thought...

So I'm trying to set up tsconfig so tsc finds '.ts' files that are in the host so it can type-check correctly. But I'm not telling mcrun about those files so it doesn't include them into the mod archive. Sounds easy....

Well, tsc has a notion of a rootDir: "The longest common path of all non-declaration input files.". The rootDir impacts the output file names: when telling tsc to output into dir X it reconstructs the paths from the rootDir. So when it compiles `/a/b/c/d.ts` and the root dir is `/a` then it outputs into `$outDir/b/c/d.js`. Since mcrun needs to pick-up the .js files output by tsc it duplicates the calculation of the rootDir. What this means is that if tsconfig is told about a file in some distant location but mcrun is not told about that file then the two may end up computing a different rootDir with the result that mcrun doesn't find what tsc output.

The solution I'm seeing right now is to take advantage of the "non-declaration input files" specificity and to ask tsc to generate declaration files when compiling the host so they can be included in the mod's compilation so the rootDir is not affected. Phew.

**Reply by phoddie** [reply]
So, basically, you are having `tsc` automate the generation of the declaration files (`*.d.ts`) for you, and then pointing to those? The paths with `tsc` always feel messy. I'm sure there's a good reason they set to up they way they did.

**Reply by tve** [reply]
Correct. The difficulty is that tsc derives the root dir so one has to guess what it arrived at. The proper solution is a mix of feeding it `rootDirs` so it doesn't go wild up the tree and then using `declarationMap` (https://www.typescriptlang.org/tsconfig#declarationMap) to find out what it did. For now I just hardcoded the inferred root dir so I can make progress on other fronts...

**Reply by phoddie** [reply]
Understood. I still imagine the manifest + mcrun could support this with a new section (as briefly considered above). But, it seems premature to take that on, so great that you are have a path forward here to continue exploring.
